extras/gizmo/gizmo.js

```diff
@@ -0,0 +1,275 @@
+    }
+
+    _createLayer() {
+        const layerMap = this.app.scene.layers.layerIdMap;
+        if (layerMap.has(GIZMO_LAYER_ID)) {
+            this.layerGizmo = layerMap.get(GIZMO_LAYER_ID);
+        } else {
+            this.layerGizmo = new Layer({
+                id: GIZMO_LAYER_ID,
+                name: 'Gizmo',
+                clearDepthBuffer: true,
+                opaqueSortMode: SORTMODE_NONE,
+                transparentSortMode: SORTMODE_NONE
+            });
+            this.app.scene.layers.push(this.layerGizmo);
+        }
+        if (this.camera.camera.layers.indexOf(this.layerGizmo.id) === -1) {
+            this.camera.camera.layers = this.camera.camera.layers.concat(this.layerGizmo.id);
+        }
+    }
+
+    _createGizmo() {
+        this.gizmo = new Entity('gizmo');
+        this.app.root.addChild(this.gizmo);
+        this.gizmo.enabled = false;
+    }
+
+    _getSelection(x, y) {
+        const start = this.camera.camera.screenToWorld(x, y, 1);
+        const end = this.camera.camera.screenToWorld(x, y, this.camera.camera.farClip);
+        const dir = end.clone().sub(start).normalize();
+
+        const xstart = new Vec3();
+        const xdir = new Vec3();
+        const v1 = new Vec3();
+        const v2 = new Vec3();
+        const v3 = new Vec3();
+
+        const selection = [];
+        const renders = this.gizmo.findComponents('render');
+        for (let i = 0; i < renders.length; i++) {
+            const meshInstances = renders[i].meshInstances;
+            for (let j = 0; j < meshInstances.length; j++) {
+                const meshInstance = meshInstances[j];
+                const mesh = meshInstance.mesh;
+                const wtm = meshInstance.node.getWorldTransform().clone();
+                wtm.invert();
+
+                wtm.transformPoint(start, xstart);
+                wtm.transformVector(dir, xdir);
+                xdir.normalize();
+
+                const pos = [];
+                const idx = [];
+                mesh.getPositions(pos);
+                mesh.getIndices(idx);
+
+                for (let k = 0; k < idx.length; k += 3) {
+                    const i1 = idx[k];
+                    const i2 = idx[k + 1];
+                    const i3 = idx[k + 2];
+
+                    v1.set(pos[i1 * 3], pos[i1 * 3 + 1], pos[i1 * 3 + 2]);
+                    v2.set(pos[i2 * 3], pos[i2 * 3 + 1], pos[i2 * 3 + 2]);
+                    v3.set(pos[i3 * 3], pos[i3 * 3 + 1], pos[i3 * 3 + 2]);
+
+                    if (this._rayIntersectsTriangle(xstart, xdir, v1, v2, v3, tmpV1)) {
+                        selection.push(meshInstance);
+                    }
+                }
+
+            }
+        }
+        return selection;
+    }
+
+    _rayIntersectsTriangle(origin, dir, v0, v1, v2, out) {
+        const e1 = new Vec3();
+        const e2 = new Vec3();
+        const h = new Vec3();
+        const s = new Vec3();
+        const q = new Vec3();
```

Comment:
Just a note to say that it might be worth using preallocated vectors here to generate less garbage.